### DISCUSSION TITLE
Big deviations between OpenEMS and HFSS

### DISCUSSION THREAD

**Discussion by nico-arnold** [thliebig/openEMS-Project#146] [Q&A Usage] [answered]
Hello everybody,

I'm planning to use OpenEMS as a solver for my Open Source project and am currently trying to evaluate the measurements vs simulations as there are some pretty big deviations.
To have a neutral starting point, I decided to take the data from an HFSS tutorial as I know how the S parameters are going to look there. I did the simulation of a microstrip structure (1.6mm FR4 substrate of 20x40mm with a 17 µm copper trace with a width of 3mm), and here are the results I get out of HFSS:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/b2387ce7-e4be-4122-b97a-a7d2059b1c9b)

Compared to that, OpenEMS results:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/f323eed8-d5c2-4b5b-8148-5622388dfb3e)
It's visibly "smoother" than the HFSS one and also has a different form.

A colleague told me I should try to decrease the copper conductivity from 58e6 to, say, 58e3, because the now increased resistance should dampen the curve a bit, decreasing the impact of reflection along the length of the microstripe. This is where it, in my opinion, started getting strange. HFSS shows similar results:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/880e08f4-4aa2-4901-9147-e6ed81ce0ea3)

But in OpenEMS, the changes of the S11 parameter are increasing.
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/43ac9e2e-98cd-4238-b42d-42932fe34f53)

However, as you can see, the general ballpark is okay. Still, S21 is decreasing in HFSS with increasing frequency (which makes sense), whereas in OpenEMS, it's actually increasing. I think that shouldn't be. What am I doing wrong here? I played around with multiple parameters including feed and measurement shift, resolution, ... But nothing improved the results.

I would be very happy if you could help me here to validate my results as I definitely want to use OpenEMS in the project, and not only for the sake of the coffee reminder x]

best wishes from Dresden and have a great weekend,
Nico

**Comment by nico-arnold** [comment]
Nearly forgot; [here's my .m and result files.](https://datashare.tu-dresden.de/s/JRjmAfQRTR9b75F)

**Reply by thliebig** [reply]
You should not (need to) change the metal losses artificially. The openEMS s21 should not go up and certainly not go above 0. Something is clearly wrong with your model. Maybe the port is "inverted" or something. 
I will see if I can find some time to have a closer look.
At the same time I would not worry to much about the s11 differences. -20 or -30dB just mean 1% or 0.1% insertion loss. This is well within the modeling details and a measurement would yet again look very different...

**Reply by thliebig** [reply]
After just a quick look, you have no room/air above you MSL? Thus you upper PEC boundary cond. touches your MSL? 
You *really* need to add some space above your MSL for the fields guided along to have room to expand...

**Reply by thliebig** [reply]
Just in general you z-mesh and boundary cond. are not ideal. You should have a closer look at the MSL Notch-Filter Tutorial maybe... For example you should not try to have multiple cells over the metal trace thickness... This will just kill the time-step.

**Comment by toammann** [comment]
Hello Nico,
I respectfully disagree with your college :). Increasing the conductor loss helps only if you have non ideal terminations at the ends of the transmission line. The transmission line acts as an attenuator in this case to provide some isolation what improves the interstage matching. This should not be the case in the simulator. To change the material Parameters is not a good idea. 

A portion of the differences you see is due to the dispersion Model of the used Substrate. I think  HFSS Users the Djorjevic-Sarkar Model by default? Not sure.

Please find further Information
[here](https://github.com/thliebig/openEMS-Project/discussions/99)

For the increasing return loss: Not sure right now. I Need to Look at your script

Can you Upload a Touchstone file Export of your HFSS Simulation?

Regards
Tobias

**Comment by nico-arnold** [comment]
Hey @thliebig , hey @toammann ,

thanks a lot for your help until here!
I'm terribly sorry, but I just realized I uploaded the wrong file. I pushed the "correct" one ("MSL_Losses_Comp.m" and "tmp" directory) into the shared folder by now and moved the old one to a subdirectory called "old_wrong", just for completeness and other people.
Anyway, the tip with the air above the trace was good as it explains why that other simulation was running but not outputting any dump nor values at all - sure, the trace was just shorted out between the ports. No wonder it didn't work. Thanks!

For the other simulation now that I took the screenshot from - it was built on top of the "OpenEMS\matlab\examples\transmission_lines\MSL_Losses.m" example file with values changed to the ones I simulated in HFSS. I re-checked, there is enough air clearance over the trace itself (7.5 mm) and zero clearance under it because there's a PEC acting as ground.
I also checked the ports, they don't seem to be inverted as they're facing onto each other.

About the tip of my colleague, yeah, that absolutely makes sense. I wasn't sure if the 50 Ohm-Port correctly terminates the strip or if it only uses that impedance to calculate values, and also how big of an impact the measurement shift has, so I gave it a try. 
I generated an S2P file and attached it beneath the .m file (:

[Link](https://datashare.tu-dresden.de/s/JRjmAfQRTR9b75F)

As next steps I will take a look at the links you sent me and also try to re-generate the file step by step from the example file. I re-evaluated the example file, and the S11 has the correct shape there.

best wishes and have a great start into the week,
Nico

**Comment by nico-arnold** [comment]
I think I kinda found the problem - it's the substrate thickness. When increased beyond ~500, it shows places where the the S11 increases above 0 dB:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/14964559-f3f2-4bbb-8b5a-260659806d6a)

I tried to densify the mesh above the trace starting from that point to get it under control and also added more air above, but neither helped:
`mesh.z = SmoothMeshLines( [linspace(0,substrate.thickness,30) 4*lambda/unit], resolution );`

**Reply by VolkerMuehlhaus** [reply]
What model did you use for this test? 
One possible reason for such unphysical S-parameters can be large (long) ports indeed, if large field gradients induce a voltage directly into the ports voltage sensor. In that case, you better use a short port plus some length of regular metal.

**Reply by toammann** [reply]
Hey Nico,
I have also noticed that your port impedance has a significant imaginary part (around 10Ohm). This should not be case. We expect a real line impedance for a microstip. 

@VolkerMuehlhaus I can also imagine, that the field cause by the pulse traveling out of the feed is somehow distorted as it hits the microstrip with an aprupt change in direction (e.g. z to x). Maybe it has to travel down the microstrip line for a short duration until the TEM mode has setteled? However, I don´t think that this should cause a phase shift between voltage and current. So, your explanation makes more sence.

Is there a rule of thumb for this problem? The substrate thickness is usually given and cannot be changed.

@nico-arnold Your action based on Volkers comment: Increase the distance between your measurement plane and the feed (change  'MeasPlaneShift') until the imaginary part of your port impedance is negligible small.

Also your real part of the Impedance is not 50Ohm. This explains why the return loss increases with frequency. You can experiment with the 'RefImpedance' setting of calcPort.  Make sure to compare the results of HFSS and openEMS with the same reference impedance. I don´t know what is the default in HFSS. It may be either 50Ohm or the ZL of the line.

You specified a fixed impedance for normalization.
```port = calcPort(port, Sim_Path, f, 'RefImpedance', 50);```

So, if your line impedance is not 50Ohm you will see a miss match to your specified reference impedance. However, if you use the ZL determined by openEMS the transmisison line should be matched over the entire band but with a different reference impedance.

```port = calcPort(port, Sim_Path, f);```

**Reply by nico-arnold** [reply]
Hey Volker, hey Tobias,

the model I'm using currently is uploaded here:
Basically still the same as before, but with some room into the x and -x direction - this produces the ripples seen in the HFSS simulation and makes the S11 match pretty well. The S21, however, still persists to be a problem as it exceeds the 0 dB line multiple times still.
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/5d410921-eff0-47e2-a0f3-73aba2fc13e1)

I tried to change the room around the model, for example by adding some padding on the y-axis, which I removed again later, and also enlarged the x-padding; no change.
Afterward, I played with the port length. I think the length of the ports shouldn't be a problem, as it appears to me that the more I reduce the length from 5000 units, the worse the problem with the S21 gets because the 0dB-excisions keep increasing. Example of reducing the port length from 5000 to 2000:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/9cdb9d05-8f1e-4951-9b5c-6e59df374151)

However, if I enlarge MSL.port_dist, then the only thing that changes is the frequency of the ripples and the impedance "ringing"/"overshoot" (don't know how to call that - what's the correct term here?) at the start frequency of 1 GHz decreases.
Moving the measurement plane only changed the frequency of the ripples:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/47341104-4baa-42d1-b2af-4f458f20bd62)

I also tried reducing the feed shift from 10*resolution down, but that also didn't make things better.

The wrong impedance was the solution I think - I noticed I accidentally swapped the "height" and "thickness" parameters in TXLine so it was giving me strange values there. I matched it for 50 Ohms impedance (MSL.with = 470 µm) now, and results look great:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/61c4a196-e019-465e-9c09-d28a0419f5cd)
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/ee2100b5-e83c-40a4-aec1-cab24de267ed)
There is still a part of the S11 that keeps sneaking above the 0 dB. I will try to further take a look at that, but that's a big improvement already.

Furthermore, I also didn't know that it was possible to automatically adapt the impedance instead of giving fixed values like I did here - thanks for the tip! Will try to play around with that.

best wishes and have a great weekend,
Nico

**Reply by VolkerMuehlhaus** [reply]
Hi Nico, 
can you upload the latest model file that you used? I want to have a look at the S11>0dB effect.
Best regards and have a good weekend,
Volker

PS: 
Based on your initial model above with 3mm line width, the CSXCAD view below might help to understand the modelling mistake. You have used 5mm reference plane shift, which is measured from the simulation boundary.
You have also specified a feed_shift to move the excitation away from the boundary.
However, both values combined place your measurement plane = voltage sensors very close to the excitation, resulting in strong fringing fields at the port 1 voltage sensors. You really need more distance there! With a larger distance between excitation and measurement plane, the error decreases by a large amount.

You noticed a dependency from line width - that is an indirect effect because the fringing fields are even worse here with the large 3mm line width.

![openems_port_too_close](https://github.com/thliebig/openEMS-Project/assets/47922981/26e366ef-48a8-4042-8110-0303b606b50c)

**Reply by toammann** [reply]
Hello Nico, 
in your latest plot has still an imaginary part of up to 5Ohm. The strong fluctuations in the real part are also not normal. I think something is still not right.

Can you increase the measurement plane shift by a significant amount? Maybe 5000um. 

You can also have a look at the measured line impedance of port 2. If the imaginary part of port 2 is zero, there is a good chance that increasing the measurement plane might help.

 I think (hope) that this should help to bring the imaginary part towards zero.

Regards,

**Comment by nico-arnold** [comment]
Hello everyone,

sorry for taking so long. There are several cases I tried; to make ientification easier, I will enumerate them.

@VolkerMuehlhaus That's interesting, didn't know that. Thanks for the tip! I thought it might be okay as it was done like that in [this OpenEMS example](https://github.com/thliebig/openEMS/blob/master/matlab/examples/transmission_lines/MSL_Losses.m). However, moving the feed indeed gave a slight improvement. I moved it all the way to the beginning of the port now (by removing the 'feed_shift' parameter): 

## Case 1
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/d6f829c1-a39d-4c79-8e48-442c21d07a2f)
Still there's a imaginary part on the impedance and also the values of S21 are not multiple dB into the negative, but stay around the 0 dB line.

You can find that model here: [https://datashare.tu-dresden.de/s/JycX7ejx53mqwzT](https://datashare.tu-dresden.de/s/JycX7ejx53mqwzT)

## Case 2
@toammann Sadly, moving the probes away from the feed (feed_shift is 10 mm now) makes the situation worse again:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/d9aa7ace-01a3-4ce5-9f17-144321e07ccd)
It's interesting to see a phase shift in the fluctuations of the Real part of the line impedance Though caused by the shift of the measurement planes. 

## Case 3
For the ports impedance, I tried removing the 'feed_R' parameters, but that leaves the simulation running forever because of the wave travelling along the strip getting reflected back and forth without actual decay.
Removing just the reference impedance on the calc_ports function gives the following results:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/9ef4c3a2-10ef-4364-9bd5-a34d96c0ef38)

So as you see, no more reflections from the ends, much smoother curve on the S11 (maybe even too smooth? especially when compared to the HFSS results). But the imaginary part of the impedance still persists. But... S21 stays well unter the 0 dB line all the time! Even if it might be too small and should be a few mor dB into the negative.
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/e8fcc245-95b5-42ee-979a-ac022ca01489)

(Reminder: This is all for a line terminated by 50 Ohm at each end with the feed and so also the impeance sitting at the diret end. Ports are 5mm off the ends at a overall line length of 40mm)

## Case 4
I also tried to decrease the impedance values (2x feed_R of ports, 1x calc_ports) to 45 ohms as that's what the simulation suggests. Now it looks as follows:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/b4d137a9-3e72-48a8-af06-a941fd86000f)
Impedance-wise, that looks pretty well right now. Note the small "ringing" at the beginning of the Z_analytic plot. From the cases before, I would have thought that by removing the fixed impedance from calc_ports should give no values above 0 dB for S21. And from the better impedance match at the transmission line of 45 Ohms I would have thought the imaginary part of the line impedance should stay around 0; seems to be the case. Also, note that the S11 curve looks again like the one before with 50 Ohm ports and no impedance provided as reference to calc_ports (case 3):
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/800dcfc3-466b-4709-89bc-9a159d07f6eb)
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/8a4ef449-34dc-4b45-b060-c66845c10820)
Also the impedance plot looks pretty much different from HFSS again, but also more like a "real" one. 

I uploaded this case here: [https://datashare.tu-dresden.de/s/LtsKtYdEMGSRbJk](https://datashare.tu-dresden.de/s/LtsKtYdEMGSRbJk)

## Case 5
To evaluate that, I checked against that case also in HFSS:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/5a62426a-9463-4709-b464-0d2d6e4b1b17)
Getting there. The shape looks more similar, maybe the rest is more or less fine-tweaking of the port length. I noticed that the deep cuts from reflections at both ends are much less defined now, though not as deep as in the OpenEMS model in case 4. Also, the lowest point of S21 is around -3 dB here while the OpenEMS model never goes even below -0.5 dB.

@VolkerMuehlhaus you said that the problem is the stringing fields at the probes caused by the broad stripe, and making it thinner reduces that error here - so it's not because of the better matcing between the feed impedance(s) and the line impedance?
Is there a way to make an automatic termination of the microstrip at both ends with a fitting impedance, or does that one always needs to be adjusted manually like I did here?

Ah, and is it better/easier for you if I upload and link a model for each separate case in the future?

best wishes and have a great week,
Nico

**Reply by VolkerMuehlhaus** [reply]
Hi Nico,

yes, it will make things a lot easier if you always upload the model for each testcase. I am working in EM support of commercial FDTD tools for many years now, and it always helps to reproduce customers issues with the exact model that they use. The issue is often in small details that they didn't even mention.

I will now look at your model. In the testcase that I had modified with more spacing between excitation and measurement plane, the resulting Zline looks fine. Model is attached.

Best regards
Volker

![openems_zline_calculated](https://github.com/thliebig/openEMS-Project/assets/47922981/2d5d33f5-e26a-445e-8c71-69ab5c028a92)

[MSL_Losses_Comp_feeddist15mm.zip](https://github.com/thliebig/openEMS-Project/files/13474483/MSL_Losses_Comp_feeddist15mm.zip)

**Reply by VolkerMuehlhaus** [reply, marked as answer]
Hello again Nico,

I looked at your model file and found a few issues and misunderstandings. For example, you try to calculate S12 and S22 without running a simulation where port 2 (and ONLY port 2) is excited. 

Another misunderstanding seems to be line length in this example: the line length parameter should be the "net" length of your device under test, excluding additional length added for the feed. The "trick" here is to shift the port reference plane (=measurement plane, the location of the port's voltage probe) to the interface where the DUT line length starts/ends. In your model, you messed up DUT line length by subtracting port_dist. 

Regarding the MUR boundary in your model: I had poor results with MUR boundary when I tested openEMS, it sometimes didn't converge and often enough I had strange results. My conclusion for my work was that PML are more reliable, so I only use that for absorbing walls in openEMS.

My proposal is that you start all over again, based on the lossy line example. I have already modified that example for your line dimensions, see attachment. This now runs with meaningful S21 values. 

[MSL_Losses_start_from_example_Nico_linedimensions.zip](https://github.com/thliebig/openEMS-Project/files/13476102/MSL_Losses_start_from_example_Nico_linedimensions.zip)

For S11, you need to consider that this model runs the line **into** the PML boundary, which means that there is **never any reflection** from the end of the line at port 2. What you see for S11 reflection is only the jump in impedance from the 50 Ohm port to the line impedance. You could use that to calculate Zline directly from Zin at port 1.

Best regards
Volker

PS: Take my description above with a grain of salt. I am very experienced with commercial EM solvers (FDTD and other methods) but quite new to openEMS, I only did some openEMS benchmarking and model development for [this ](https://github.com/IHP-GmbH/IHP-Open-PDK/tree/main/ihp-sg13g2/libs.tech/openems)project.

**Comment by nico-arnold** [comment]
Hello @VolkerMuehlhaus ,

thanks a lot for your help! It indeed solved many points. I will answer back thoroughly next week as I'm on a road trip right now (:

best wishes,
Nico

**Reply by nico-arnold** [reply]
Hello @VolkerMuehlhaus ,

so finally some time to reply here, a comfy chair and a desktop instead of my car and old laptop...

I ran your examples and indeed, it works flawlessly now. Thanks a lot! What wonders me about the example with the transmission line length being the "net length" without the port length - isn't it done just the same way as in my file in the "CPW_Line.m" example in the OpenEMS/matlab/examples folder?

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/732af924-eacf-496c-ab75-8a9571a9856a)

About S11 and S21: good point indeed, I didn't know that. But now that you said it, it makes perfectly sense and I don't know why I did not think about that. Anyway, I didn't visualize those, so it should be no problem that those values are there, I think.

Concerning the "flat" curves in OpenEMS of course you're also right. I totally forgot that HFSS might use different boundaries and automatically assumed it would also be something similar to PMLs. Indeed, if you "smooth out" the HFSS curves, you get nearly the same curve as in OpenEMS. At least for S11. S21 seems to be a bit smaller in OpenEMS (not even -0.5 dB opposed to roughly -2 dB in HFSS). But I think that's all in the correct ballpark here.
So after adding PMLs to HFSS it should give the same curves. I will make a test again during the next days.

So after all, I'm working without specification of the feed/measurement impedance now, starting from your model. I'm trying to make a simple CPW now and continue to a grounded design (CPWG) and afterwards differential geometry from there. The impedance value seems to be in the right size (90+ Ohms opposed to 106 Ohms from TXLine), but I still struggle with the reactance going down to -20 Ohms. I will try to use what I learned here to solve that problem by myself; if I can't, I hope it's okay to ask you again next week? (:

best wishes, have a great weekend and a nice 3rd advent,
Nico

**Reply by VolkerMuehlhaus** [reply]
Hi Nico,

> What wonders me about the example with the transmission line length being the "net length" without the port length - isn't it done just the same way as in my file in the "CPW_Line.m" example in the OpenEMS/matlab/examples folder?

The difference is that your model subtracted port offset length here: 
`MSL.length = 40000-2*MSL.port_dist;`

But the remaining code in the model applies some offset **again** and places the effective reference planes at x=-10mm and x= +10mm, for a total line length (as extracted in S-parameters) of 20mm. You can check that in AppCSXCAD, move the grid plane along the x axis to read out x position.

The original code in the example just assigns a value to MSL.length, without subtracting the port distance in **that** code line. Check both models in 3D geometry preview, with ProbeBox items enabled, and you will see what I mean.

> S21 seems to be a bit smaller in OpenEMS (not even -0.5 dB opposed to roughly -2 dB in HFSS). But I think that's all in the correct ballpark here.

Not sure what you modelled in HFSS for the substrate, but that might be from dielectric losses which are not included in this openEMS model. There is [another discussion](https://github.com/thliebig/openEMS-Project/discussions/157) going on these days how to model wideband dielectric losses in openEMS.

Have a great weekend!
Volker

**Reply by VolkerMuehlhaus** [reply]
Let me add that openEMS gives really accurate results, if you know the limitations of each model. Below is a comparison of openEMS results for an RFIC line in SiGe technology measured up to 110 GHz. Not so easy to measure that accurately to within 0.x dB insertion loss at those frequencies...

The "trick" required in that model was to mesh into skin effect, because thick conductor modelling was required (line width 15um at 3um metal thickness) and openEMS does **not** have a **thick** conductor model with built-in skin loss calculation. The solution was to use the generic material object and mesh into skin effect. So we sometimes need some tricks, but results are fine in the end...

![openEMS_vs_measured_SG13G2](https://github.com/thliebig/openEMS-Project/assets/47922981/2d6c57bd-78dc-4eb6-8bb9-15be55bacee4)

The model is published in [another project](https://github.com/IHP-GmbH/IHP-Open-PDK/tree/main/ihp-sg13g2/libs.tech/openems/testcase/SG13_line).

**Reply by nico-arnold** [reply]
@VolkerMuehlhaus 
Yes, I get your point, but that value is only subtracted once, so it's the same as writing down an overall length of 30 mm for the transmission line instead of 40. I mean, shouldn't the impedance stay (nearly?) the same there?

Indeed, those charts are impressive, that's pretty precise. At least if you know how to design the FDTD simulation (hehe), which is something that I still have to learn more. Do you have literature recommendations additionally to the ones mentioned [here](https://docs.openems.de/more.html)? I will try to get them in the university library tomorrow to gain a bit more knowledge.

The lossy substrate model is also definitely interesting, thanks for pointing me there - It would also explain the drop of the S parameters with increasing frequency in HFSS being stronger than in OpenEMS. Indeed, I only specified the dielectric constant here, but not the loss tangent.


I think I have the simulation for the CPW running now - or, at least, the values appear to make sense to me and seem to respond to what TXLine recommends for the given geometry. Tomorrow, I will try to rebuild this one for a grounded CPW (CPWG) and see if I get also matching results.

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/ebbae377-42ee-41ad-bd8b-4def3abbecb7)
I get a rise in impedance towards DC, which should be correct - with the line being unterminated, it should show an open circuit or DC.

Are there any examples for differential ports? I could not find some hints in the examples and the Matlab files like AddCPWPort. I need to verify a design for a differential CPWG, as I have measurements for that one to validate my model with. The only hint I have right now is what is generated by [PyEMS](https://github.com/matthuszagh/pyems).

Matlab/Octave files are [uploaded here](https://datashare.tu-dresden.de/s/4K4s3fegtk9dyHd). The correct one is called _CPWTest_PML8.m._ I also had a look how everything behaves when replacing the -z boundary with a PEC instead of PML_8, but interestingly, it only shorts out my line impedance for frequencies towards DC. I'll have a closer look tomorrow.

**Reply by VolkerMuehlhaus** [reply]
> Yes, I get your point, but that value is only subtracted once, so it's the same as writing down an overall length of 30 mm for the transmission line instead of 40. I mean, shouldn't the impedance stay (nearly?) the same there?

As I wrote above, the reference plane in your code shift will again change the effective length that you see in results. You can do whatever you want, as long as the "net" length of line between the reference planes is the length what you expected to simulate. 
For your very first line with excessive width, the net line length was getting rather short after increasing the port reference shifts, because of your code modification. I would suggest to keep the net line length a truely independent parameter, and add possible reference plane shifts on top of that. 

One side effect of the shift that you applied is that all objects, including substate and ground, where then move away from the simulation boundary, whereas they touched the boundary in the initial post. Just check in AppCSXCAD that you get what you expected to model, that's all I am trying to say.

Literature: I don't know what literature to recommend here, my introduction to these topics was from manuals and seminars for the commercial EM solvers. The other users might have some ideas, hopefully.

CPW port: I have no experience with using that in openEMS. I suggest that you study the internal structure of the CPW port in AppCSXCAD, to see where the source and voltage probes and terminations are. That might give an idea where you can add vias to use the port for feeding GCPW.

Regarding your short, you can check the distance to top boundary, and place an additional mesh line at larger distance to push the PEC away. EDIT: I misunderstood your setup, and need to look in more detail what is going on here.

**Reply by VolkerMuehlhaus** [reply]
Update: For unknow reason I have difficulties to run openEMS models on Octave, something seems to die after a while, and any output to the command window stops. 

I could not find out what exactly is causing the high S11 at DC, but this looks like an artefact to me. It doesn't really make physical sense to me when running the line into a PML = termination. S-parameters in FDTD are a fourier transform of the time domain simulation results, and low frequency values (long period) would require time domain covering a really long time. Looking at the time column in the output file, we see that FDTD simulation stopped at 47ns when the -40dB energy threshold was satisfied, so that is not enough time to precisely extract the low frequency response or DC. It can work, but it's not guaranteed, as we can see here.

Other topic: The model that you created looks a bit overkill to me, with all the air volume meshed at 30 cells per wavelength, and the PML at that large distance. It should work, but will need rather long simulation time for such a trivial model. For models that don't radiate, we don't need that much distance to the sidewalls.

**Reply by nico-arnold** [reply]
Hello Volker,

ah, okay, good point. I think now I understood the point, I was a bit short-sighted there. Thanks for the clarification!

Ah okay, I didn't know it shouldn't be infinite for DC, I expected it to just behave like an open circuit there. But you could be right. I tried to narrow that down by simulating across a smaller frequency range (0-20 MHz instead of 0-20 GHz), but because of the frequency dependence of the resolution this fails to mesh. Because of the new f_max being 1/1000th of the one before, I also divided the resolution by 1000 to have the same ration, but that seems to give completely different results. Also, I reduced the airspace in any direction to 1000 instead of 7000 to shorten the simulation times. This took ages to even start the FDTD engine, so I reduced this to 100. Sadly, this gave totally different results.

0-20 GHz:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/d5ba6403-2a26-4e10-a58d-d46ff5084802)

0-20 MHz:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/e84208ad-aa6c-46d0-9eb7-fd51c4ec4833)

[Octave files](https://datashare.tu-dresden.de/s/qkNAtXcBDJP9q8n)

I'm going to have a closer look at that tomorrow, also by reading some more fundamental theory and reading some posts on the discussions here. But that just as a short side notice, I think tomorrow I will see there's just a simple mistake that I made somewhere.

For the GCPW, yes, that was exactly my plan: I think I should be able to derive a _AddGCPWPort.m_ from _AddCPWPort.m_ and _AddMSLPort.m_.

Off to the library now, fetching some lecture to make a deeper dive into it (:

best wishes,
Nico

**Reply by VolkerMuehlhaus** [reply]
Yeah, getting some theory on FDTD sounds like a good plan! This will also help to understand why low frequency is rather difficult, especially if the physical model dimensions are small.

When I looked at your model yesterday, I couldn't figure out why it breaks at DC. But I noticed that the CPW tutorial example uses port resistors instead of placing the port at the absorbing boundary, and that simulated with no such DC issue. Overall, port with "real" discrete resistors seems the better model anyway, because you will get the expected reflections from ports that are not matched to the line impedance. 
There are always some small port parasitics, so don't expect to see -60dB return loss as you do for the line into PML boundary, but overall these ports should perform fine.

Best regards
Volker

**Reply by nico-arnold** [reply]
Hello Volker - have a happy new year! (:

Small update from me. I have 3 books and an [eBook](https://eecs.wsu.edu/~schneidj/ufdtd/) here now about FDTD. So far, I like the eBook the most as it appears more engineering-directed, better understandable and not mostly about the mathematical background of the FDTD method and fields. I'm not too good with the latter one, sadly.

Good point, that makes sense indeed. So for smaller frequencies, I will have to look for a more suited method to do the simulations then. We're speaking about a range of up to a few hundreds MHz here, so that should maybe still be fine, though. As I noticed, the 20 MHz I tried to simulate are already getting out of hand.

I tried to compare my model to the [_CPW_Line.m_](https://github.com/thliebig/openEMS/blob/master/matlab/examples/transmission_lines/CPW_Line.m) official example and noticed that it seems to have a similar problem here. I changed the last few lines so that the impedance values get calculated, the rest is the same:

[Octave/Matlab file](https://datashare.tu-dresden.de/s/YmN4LSXBJHNELw8)

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/197c5c07-2aa8-409f-908b-eb8dafc7285b)

When taking a closer look, you can see that the impedance also has a bigger imaginary part:

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/88daa7fa-45e2-430e-9ccf-7eb95b2edbb2)

Closeup of the DC breakdown:

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/00c4f251-0bbc-489b-9de9-e417988670a7)

That appears a bit weird to me. I will reinstall OpenEMS later and try to find out why I have that dip around DC and you don't. Maybe I messed some things up while playing around.

best wishes,
Nico

PS: I'm thinking about changing the title to "Comparing OpenEMS and HFSS results" as it appears more fitting and not like "something's terribly wrong". Is that okay for you?

**Reply by VolkerMuehlhaus** [reply]
Hi Nico,

happy new year to you!

The ebook that you found looks interesting, I have saved the link for future use.

Regarding the plot with ripple and imanginary part, that is exepcted because you now simulate the "real" S-parameters of the line which is not perfectly matched, terminated into 50 Ohm port, and your code line 

`ZL = port{1}.uf.tot./port{1}.if.tot;`

gives the **input impedance into port 1** of the line. Now that we have terminated the line into a 50 Ohm port instead of running it into a PML, this value is no longer the transmission line impedance Zline=sqrt(L'/C') !! The input impedance circles around ~ 52 Ohm in Smith chart, resulting in the very typical pattern that you showed.

Regarding low frequency: yes, strictly speaking we can't get the true low frequency (or DC) response from the very limited time period that we evaluate after excitation. The commercial time domain solvers have a feature called "resonance estimate" (Empire XPU) or "AR filter" (Microwave Studio) for better extrapolation down to DC assuming that there are no unexpected resonances at very low frequency. But even without that feature, many of my openEMS models had a smooth response towards DC, so I'm not sure what is different here.

Regarding the thread title, no worries, do whatever you consider best. It is very common that users need some time to adjust to a different solver method, and all the methods have their specific strengths and weaknesses and gotchas.

Best regards
Volker

**Reply by nico-arnold** [reply]
Hey Volker,

thanks! (:

Yeah, that book is a good lecture. There'a also a link on the page where you can download the whole PDF.


Indeed, the fact with the port resistances also hit me yesterday. I had tried to remove them before, but the simulation ran much longer. Only after submitting my last comment, I realized that it was running forever because the line wasn't ending in a PML, so there is no way to decay the excitation with the wave getting reflected back and forth. Stupid me....
In the evening, I had a small test at home that I also just re-run on the bigger computer at work, with the same results. I took the above example again and just changed the `if 1` in line 30 to `if 0` so it was running the "else" part there, removing the port resistances and also the airspace in x-direction, making the line end in the PML. What I expected was
- less reflections leading to a more smooth response
- no imaginary part on the line impedance due to a perfect match
But, instead, I get this result:
![image](https://github.com/thliebig/openEMS-Project/assets/82146516/11cfebf5-affe-4591-bc1b-d79aa3c2161c)
I verified this in ParaView, and indeed, you can see the wave get reflected at the end of the CPW and travel back. But why's that? In the given example, in the line `pml_add_cells = [0 0 8 8 8 8]; ` it is specified that there should be no clearance to the PML in x-direction. I also tried to add 8 cells of space in the x-direction there, but that didn't make things better.
I re-checked everything, but it doesn't make sense to me. PMLs to decay waves? Check. No problems with port length? I think so, it's the official example. 

[Octave file](https://datashare.tu-dresden.de/s/mq73ip6fEBpXC9R)

**Reply by nico-arnold** [reply]
Okay, my mistake. The BCs are defined as [2 2 2 2 2 2] which is MUR - changing all to 3 (PML) resolves the problem:

![image](https://github.com/thliebig/openEMS-Project/assets/82146516/d3b575dc-64e7-4c78-84e4-1d94a1269bd0)
